sweep/sweeper.go

```diff
@@ -687,22 +687,57 @@ func (s *UtxoSweeper) removeExclusiveGroup(group uint64) {
 func (s *UtxoSweeper) sweepCluster(cluster inputCluster,
 	currentHeight int32) error {
 
-	// Execute the sweep within a coin select lock. Otherwise the coins that
-	// we are going to spend may be selected for other transactions like
-	// funding of a channel.
+	// Execute the sweep within a coin select lock. Otherwise the coins
+	// that we are going to spend may be selected for other transactions
+	// like funding of a channel.
 	return s.cfg.Wallet.WithCoinSelectLock(func() error {
-		// Examine pending inputs and try to construct
-		// lists of inputs.
-		inputLists, err := s.getInputLists(cluster, currentHeight)
+		// Examine pending inputs and try to construct lists of inputs.
+		allSets, newSets, err := s.getInputLists(cluster, currentHeight)
```

Comment:
Hmm I am a bit undecided whether it's worth it introducing this short term fix in this way, why:

Let's pretend we do not have a node restart yet (because this marks every not swept input as new again), we are currently not really verifying the RBF constraints, this means we could easily bundle new sweeps with retried one(already in mempool) and not fullfill the BIP125 requirements. 

What I propose instead would be:

1. Exclude the AnchorOuputs in a different input set (the anchors sweeps which are swept at min realy fee) => because third party sweeps which bundle a significant amount of anchors can brings us in negative-yielding territory very fast, therefore its important to exclude them into a different sweep set.
2. I would switch the order meaning that I would first sweep the new inputs first. Because we do not verify the RBF conditions yet, we should not risk bundle an input with an already published one, especially until 1) is not fixed, but also imagine fees go done, we would bundle the new inputs with retired one (published when fees were higher) reducing the successful broadcasting of the tx to almost zero.
3. After every restart of lnd we should definitely check for all inputs and whether they are already in our mempool or even spent (using the `gettxout` rpc call) because rescanning can take a lot of time I experienced on my system. When we have a clear view which Inputs are already in our mempool, we still have a clear view between new inputs and retried ones. 

So to sum it up, I think we need to take care of 1) first. Until we have a clear picture of whats happening in our mempool and how many inputs are already sitting there at which feerate, we should broadcast the newsets first (even don't mix retired inputs with new inputs yet).

Reply:
So the general direction of refactoring sweeper is to remove the contextual info such as anchors or second level HTLCs, and focusing only on creating/publishing transactions **economically**. And all the contextual info should be abstracted and specified via the `Params` field. I don't know this is possible or not, still digging. That being said, the sweeper shouldn't care whether this is an anchor input or other kinds, you just tell it how you wanna it being swept via the `Params` field.

As for the order of new inputs vs all inputs, tbh I'm not sure here - my thought is we should try to sweep all inputs first, and if there's an error, we then sweep the new inputs, otherwise the old inputs would just be sitting there. There's also a chance that the RBF condition will be met when we sweep all inputs, for instance the mempool conditions changed and the fee estimator now gives a higher fee rate - of course we don't wanna relay on this uncontrolled behavior, which is something we'll fix in the following PR to handle the RBF case properly.

Reply:
> As for the order of new inputs vs all inputs, tbh I'm not sure here - my thought is we should try to sweep all inputs first, and if there's an error, we then sweep the new inputs, otherwise the old inputs would just be sitting there.

That sounds good to me as well, but with the current change we would not do this or, at least not until all retried inputs reached their maximum retry attempt(default=15)? By that time we could in the worst case have to wait quite a bit until the conflicting inputs are gone ? The code before this PR would at least try them as well, so my concern is that we are not temporarily (until follow-up PR's come in)  improving the sweeper code with this change or but I might miss something here ?

Reply:
About the anchor inputs: I also think that placing anchors in the same tx as other sweeps may introduce extra complexity but also higher fees overall. Maybe they could have their own `ExclusiveGroup` which is anchors-only :+1: .

The rationale for this would be that anchors are lower-priority sweeps of inferior significance. If there is an on-going sweep attempt for inputs of high significance, adding anchor inputs is gonna increase the total weight of the tx which is meant to be swept fast (maybe because of other inputs' timeouts).